### PR TITLE
[decision]: Changetracking (round 2)

### PR DESCRIPTION
**Pull request by atmaxinger**
## Basics

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

**Comment by atmaxinger** [comment]
I've mainly updates some alternatives with regards to full copy-on-write within Elektra. 

One more thing that we need to discuss, is what we do with false positives - i.e. keys that are detected as changed even though practically they aren't. This can (and will) happen with plugins that do transformation of the key values. 

As of now, I've added to the assumptions that false positives may be okay. We might be able to provide some advanced API that plugins can even filter out false positives.

**Comment by markus2330** [comment]
> One more thing that we need to discuss, is what we do with false positives - i.e. keys that are detected as changed even though practically they aren't. This can (and will) happen with plugins that do transformation of the key values.

Please elaborate (ideally within the decision, maybe in notes). If I change a value and then back it is obviously not the same reference (needs more memory) but a string-comparison still shows it is unchanged. Doesn't the comparison avoid any false positives?

**Comment by atmaxinger** [comment]
@markus2330 It is elaborated in the decision, under "Assumptions". It has nothing to do on how we compare values, but that plugins may change the values before writing to disk.

**Comment by markus2330** [comment]
Thx, I see! Couldn't this problem solved as well if we require users to call `kdbSet` after any changes in the KeySet and the transformation happens there? For this decision I would simply assume that transformations are done correctly and discuss how the transformations should work in the transformations decision. #4621

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (markus2330) [inline doc/decisions/1_in_discussion/change_tracking.md:69]:
```suggestion
### Alternative 1 - Tracking within `libelektra-kdb`
```

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (markus2330) [inline doc/decisions/1_in_discussion/change_tracking.md:73]:
```suggestion
Essentially, we'd have an internal cache for all keys that were used returned by a `kdbGet` operation in the lifetime of the `kdb` instance.
```

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (markus2330) [inline doc/decisions/1_in_discussion/change_tracking.md:142]: `backendData->keys` is an implementation detail. Is it the internal cache or something else?

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (markus2330) [inline doc/decisions/1_in_discussion/change_tracking.md:148]: Again, please avoid internals like `backendData->keys` (or introduce them, as not everyone knows the details of new-backend yet).

**Review by markus2330 — CHANGES_REQUESTED** [review]
Good to see progress here!

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (atmaxinger) [reply doc/decisions/1_in_discussion/change_tracking.md:142]: If we leave out this detail, then yes it would probably be the same as Alternative 1. This distinction was made before copy-on-write, as `backendData->keys` already stores the keys returned, whereas for Alternative 1, we'd have to have another duplication.

With copy-on-write, this distinction gets much more fuzzy.

**Comment by atmaxinger** [comment]
> Couldn't this problem solved as well if we require users to call kdbSet after any changes in the KeySet and the transformation happens there?

I don't see how this would solve anything. The big problem is that something _other than the user_ can change the values.

This means that plugins using the change tracking API after the transformation step (e.g. notifications plugins and in particular the record plugin) will potentially get other change tracking results than plugins before the step.

From my POV, the "truest" change tracking results could be determined by:

- Storing a copy of the keys in `kdbGet` directly after the keys have been read from disk. Before any value transformations have been applied but after keyname transformations have been done
- Comparing it with the keys in `kdbSet` after everything is written to disk, but we somehow still have access to the keyname before the transformation.

This way, we would only detect if something truly changed. But this also means change tracking would be useless for all plugins other than notification hooks and the record hook. Those are the only plugins that are executed AFTER everything is done.

**Comment by kodebach** [comment]
> From my POV, the "truest" change tracking results could be determined by:

That goes in the direction of what I suggested in #3497 a long time ago. But doing such a split on the `libelektra-kdb` level would be another (smaller) new-backend change.

I actually think, the "true" diff would be calculated between what is returned by `kdbGet` and what is passed to `kdbSet`. That's the data that is visible to applications, that is "Elektra's view of the world". The whole idea is that Elektra gives you a uniform view of the config, no matter what is actually stored, so that's what we should use.

> This means that plugins using the change tracking API after the transformation step (e.g. notifications plugins and in particular the record plugin) will potentially get other change tracking results than plugins before the step.

I don't quite understand that statement. I'd assume the diff is calculated once in `kdbSet` and stored as long as `kdbSet` is running. If a plugin requests the diff via the API, it just gets this one diff, no matter when the plugin is executed.

We could of course also provide an API for calculating diffs (should be a separate library though). But that API should IMO take 2 `KeySet`s as input and produce a diff. So if plugins used that API for some reason, they'd have to provide both `KeySet`s. Then it is obvious why plugins get different results, if the have different `KeySet`s

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (kodebach) [inline doc/decisions/1_in_discussion/change_tracking.md:63]: Let's use something that really exists

```suggestion
    Scenario: `type` plugin configured for boolean conversion with `false`/`true`
    - `system:/background` is stored with value `false`
    - user gets key `system:/background` with value `0` (after conversion by `type`)
    - user changes it to `false`
    - changetracking detects that value has been changed, because `false` != `0`
    - `type` plugin changes value to `0`
    - consumers of the changetracking API will now get a false positive if they query whether `system:/background` has been changed
```

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (kodebach) [reply doc/decisions/1_in_discussion/change_tracking.md:142]:
```suggestion
Currently, however, this is not a deep copy, as we are returning the internally stored `Key` instances directly.
```

We don't need to say where they are stored, just that the keys are returned and therefore can be changed externally.

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (kodebach) [reply doc/decisions/1_in_discussion/change_tracking.md:148]:
```suggestion
Another problem with this approach is that the internally stored keys are recreated as new instances every time `kdbGet` is called.
```

Not quite true I think, but good enough here.

**Review by kodebach — COMMENTED** [review]
I do agree with markus2330 that we should assume transformations are done correctly (whatever that may mean) in this decision. At least for now. If we find that there is no correct way to do them, then we may need to revisit this.

However, I also think false positives **for values** are not an issue at all. False negatives would be a much bigger issue, because that means we miss changes. If there is a use case where false positives must be avoided, we can just do a kind of re-check.

The change tracking plugin should have access to the current data. It can just check if the diff matches the current data and if not update the diff.

That said, I also believe that in your example the user using `red` (or in my example `false`) can simply be called incorrect. Then the reason for the false positive is simply that the user made an incorrect change. I say "incorrect", because the user should use the format returned by `kdbGet`. So if `RED` is returned then they should use `RED` and not `red`. In fact, you could even argue that the plugin that does the uppercasing should raise an error during `kdbSet`, if it sees a non-uppercase value.

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (atmaxinger) [reply doc/decisions/1_in_discussion/change_tracking.md:63]: Funny, in #4621 it was suggested that we don't name a plugin specifically 😅

Comment (kodebach) [reply doc/decisions/1_in_discussion/change_tracking.md:63]: Hm, I guess there are pros and cons. I just think giving it a name makes this example more readable. I wouldn't even have to be a specific existing plugin. Also I think the `false` -> `0` conversion would be much more common and therefore it would be easier to understand why this is useful/important.

So maybe we go with:

```suggestion
    Scenario: plugin that converts `false`<->`0` and `true`<->`1`
    - `system:/background` is stored with value `false`
    - user gets key `system:/background` with value `0` (after conversion by plugin)
    - user changes it to `false`
    - changetracking detects that value has been changed, because `false` != `0`
    - plugin changes value from `false` to `0`
    - consumers of the changetracking API will now get a false positive if they query whether `system:/background` has been changed
```

Note also the additional line at the beginning explaining how the value is stored.

**Comment by markus2330** [comment]
Please rerequest the review when I should have a look.

**Comment by atmaxinger** [comment]
I took the liberty and restructured this decision a bit. I have found that there are 3 largely independent things that we need to decide on here:

- Where to store changetracking data
- Where to implement changetracking
- How plugins and applications can interact with the changetracking mechanism.

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (kodebach) [inline doc/decisions/1_in_discussion/change_tracking.md:57]:
```suggestion
  - this may happend when some keys have been changed by the user, but have subsequentially been "unchanged" by a transformation plugin
```

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (kodebach) [inline doc/decisions/1_in_discussion/change_tracking.md:64]:
```suggestion
    - consumers of the changetracking API will now get a false positive if they query whether `system:/background` has been changed
    - We assume consumers of the changetracking API have access to both the previous and the new value of a changed key.
      Therefore, they could detect these false positive cases, if the need to.
```

**Review by kodebach — COMMENTED** [review]
LGTM, but if you think the three parts are independent, it might make sense to split them into 3 decision files as well.

I'm not 100% sure it makes sense, since they are still quite related and share constraints, assumptions and most of the problem. But it may be easier to move forward, if we can decide the parts independently.

**Comment by atmaxinger** [comment]
Well, I don't think they are 100% independent. What I mean is that those are 3 areas we need to decide on, but it's still a single decision. Otherwise, I would have to repeat some stuff in each alternative (e.g., if the implementation is in a plugin, then it could still be possible to use `backends->keys` but also plugin-internal storage)

I just want to present some alternatives for each area, and then we pick the most suitable ones from each.

**Comment by atmaxinger** [comment]
jenkins build libelektra please

**Comment by atmaxinger** [comment]
@markus2330 could you please look over the decision?

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (markus2330) [inline doc/decisions/1_in_discussion/change_tracking.md:160]: Please add some context why this would be needed (connect to use case etc.). Isn't three KeySets with all added/removed/modified (meta)keys enough?

**Review by markus2330 — APPROVED** [review]
Definitely looks fine for in_discussion. Before it can be in solutions_clear please add how a plugin would provide "Solutions - Query".

File: doc/decisions/1_in_discussion/change_tracking.md
Comment (atmaxinger) [reply doc/decisions/1_in_discussion/change_tracking.md:160]: Good catch! Yes, if `elektraChangeTrackingGetModifiedKeys` / `elektraChangeTrackingGetModifiedMetaKeys` return the *old* keys, then these functions are not required.

**Comment by atmaxinger** [comment]
@markus2330 I have added an entry as to how the query mechanism could be implemented in a plugin. Do note, however, that I actually intended to only have the API within a library (i.e. `libelektra-kdb`). This API wouldn't care whether changetracking is implemented as a plugin or not.

**Comment by markus2330** [comment]
I think we can merge this but there are build errors.

**Comment by atmaxinger** [comment]
The build error is a timeout in `npm install`. I triggered a rerun of the failed job.

**Comment by atmaxinger** [comment]
jenkins build libelektra please

**Comment by atmaxinger** [comment]
@markus2330 @kodebach CI is green, we can merge now.

**Pull request merged**
